### PR TITLE
Svelte: Support v4

### PR DESCRIPTION
**Pull request by JReinhold**
## What I did

- Expanded the dependency ranges to support Svelte v4

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

**Comment by JReinhold** [comment]
@shilman @yannbf this should be patched back to `7.0` (once everything is confirmed to be working ofc) because of the expanded version range, but I'm not sure if the changes to templates should too, since they now install `svelte@next` instead of `latest`. Will it have an affect on the CI we run for `main`/`7.0`, the status page, or something else?

File: code/frameworks/svelte-vite/package.json [hunk 1]
```diff
@@ -52,7 +52,7 @@
     "@storybook/svelte": "7.1.0-alpha.27",
     "@sveltejs/vite-plugin-svelte": "^2.0.0",
     "magic-string": "^0.27.0",
-    "svelte": "^3.0.0",
+    "svelte": "^3.0.0 || ^4.0.0 || ^4.0.0-next.0",
```
Comment (valentinpalkovic) [inline code/frameworks/svelte-vite/package.json:55]: Package managers will install the latest version. Defining these version ranges doesn't make therefore much sense. We have two options here

1. Either we define version 4 and the latest version will always be installed
2. We move this to the peer dependency section (BREAKING CHANGE?!) and adjust the CLI to automatically install svelte for svelte-vite framework.

Comment (JReinhold) [reply code/frameworks/svelte-vite/package.json:55]: oh yeah this should be peer, it's probably a regular dependency by mistake, option 2 would be the right call.

@shilman would that be a breaking change?

Comment (ndelangen) [reply code/frameworks/svelte-vite/package.json:55]: Let's make it a peerDependency and a devDependency, similar to vite/react
It makes sense.

svelte users of storybook will have svelte as a dependency... surely?!

Comment (JReinhold) [reply code/frameworks/svelte-vite/package.json:55]: Yeah I came to the same conclusion @ndelangen, I can't think of a scenario where users wouldn't have Svelte installed already. Maybe in a specific monorepo setup where it's in another package, but I don't know.

I'd consider this more of a bug fix than a breaking change.

Comment (JReinhold) [reply code/frameworks/svelte-vite/package.json:55]: @ndelangen, `@storybook/sveltekit` depends on `@storybook/svelte-vite` (it's a thin wrapper around it), so making `svelte` a peerDep in `@storybook/svelte-vite` would mean we should do the same in `@storybook/sveltekit` right? Currently there's no dependency on `svelte` at all in `@storybook/sveltekit`.

Comment (ndelangen) [reply code/frameworks/svelte-vite/package.json]: yes, correct

**Review by kasperpeulen — APPROVED** [review]
Testing approval here

**Comment by JReinhold** [comment]
@kasperpeulen do you know why types are failing here? https://app.circleci.com/pipelines/github/storybookjs/storybook/52536/workflows/580c415a-ebab-4aa8-97e9-58cd32596d46/jobs/544399 (I can reproduce this locally)

It's failing [here](https://github.com/storybookjs/storybook/blob/svelte-4/code/frameworks/svelte-vite/src/plugins/svelte-docgen.ts#L8) because it can't find the `'svelte/compiler'` module, but I can't reproduce this outside of our Storybook repo. In a separate project, doing the same import (`import { preprocess } from 'svelte/compiler';`) and going to definition correctly takes you here: https://github.com/sveltejs/svelte/blob/master/src/compiler/preprocess/index.ts#L210

Are we configuring TypeScript in a way that doesn't support these direct-to-JS definitions, or something else?

cc @benmccann

**Comment by benmccann** [comment]
Thanks for testing! We'll release a new version in the next day or two, which should hopefully fix this

**Comment by benmccann** [comment]
`4.0.0-next.2` is now out. Can you check if it fixes it?

**Comment by dummdidumm** [comment]
The `svelte/internal` thing failing is expected - we deliberatly removed those types to discourage their use. You can still use them, but you need to silence the type error. Probably best to talk about the need to use `svelte/internal` separately in Discord. I suspect there are more projects in the ecosystem resorting to internal methods to accomplish certain tasks, and we want to make sure they are properly supported in Svelte 5.

**Review by ndelangen — APPROVED** [review]

**Comment by JReinhold** [comment]
Thanks for the detailed description @dummdidumm. We're planning to remove this area of the code soon, so I wouldn't worry too much about looking into long term support of the internal features that we use there.

File: code/renderers/svelte/package.json [hunk 2]
```diff
@@ -64,12 +64,12 @@
...
   "devDependencies": {
     "expect-type": "^0.15.0",
-    "svelte": "^3.59.1",
+    "svelte": "^4.0.0",
     "svelte-check": "^3.4.3",
     "typescript": "^5.0.4"
   },
   "peerDependencies": {
-    "svelte": "^3.1.0"
+    "svelte": "^3.59.1 || ^4.0.0"
```
[the remark this answers is not in the document]
Comment (JReinhold) [reply code/renderers/svelte/package.json]: oh no you're right, I missed that. thank you, reverted.

**Comment by shilman** [comment]
@JReinhold I don't feel great about patching this back to 7.0 since it's technically a breaking change. Before users did not have to have svelte installed, and now they do. It's a corner case, but I'd much rather have this only go out in a minor release. Are you ok with that?

**Comment by JReinhold** [comment]
> @JReinhold I don't feel great about patching this back to 7.0 since it's technically a breaking change. Before users did not have to have svelte installed, and now they do. It's a corner case, but I'd much rather have this only go out in a minor release. Are you ok with that?

We discuss this a bit here. https://github.com/storybookjs/storybook/pull/22905#discussion_r1218147216

I would consider this a bug fix more than a breaking change. And that any users that are affected by this are relying on a bug.
And for semver there's really no difference between patch and minor for most users.

But it's your call.

**Comment by shilman** [comment]
@JReinhold Ok I'm fine with that. Just wanted to discuss first

**Pull request merged**